### PR TITLE
 feat(GraphQL): RHICOMPL-3542 filter systems by rule severity

### PR DESCRIPTION
**Pull request by romanblanco**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/graphql/types/system.rb [hunk 1]
```diff
@@ -47,7 +47,12 @@ class System < Types::BaseObject
       argument :profile_id, String, 'Filter results by profile ID',
                required: false
     end
-
+    field :rules_failed_severity, [::Types::Rule], null: false do
```
Comment (skateman) [inline app/graphql/types/system.rb:50]:
```suggestion
    field :rules_failed_by_severity, [::Types::Rule], null: false do
```

File: app/graphql/types/profile.rb [hunk 2]
```diff
@@ -45,41 +45,33 @@ class Profile < Types::BaseObject
     field :external, Boolean, null: false
     field :parent_profile_id, ID, null: true
     field :in_use, Boolean, null: true
-
```
Comment (skateman) [inline app/graphql/types/profile.rb:48]: ??

File: app/graphql/types/benchmark.rb [hunk 3]
```diff
@@ -17,7 +17,6 @@ class Benchmark < Types::BaseObject
     field :profiles, [::Types::Profile], null: true
     field :rules, [::Types::Rule], null: true
     field :value_definitions, [::Types::ValueDefinition], null: true
-
```
Comment (skateman) [inline app/graphql/types/benchmark.rb:20]: ??

File: app/graphql/types/system.rb [hunk 4]
```diff
@@ -64,6 +69,15 @@ def rules_failed(args = {})
...
     end
 
+    def rules_failed_severity(args = {})
+      ::RuleResult
+        .for_policy(args[:policy_id])
+        .failed
+        .joins(:rule).where(rules: { severity: args[:severity] })
+        .joins(:host).where(hosts: { id: object.id })
+        .select('rule_results.host_id', 'rules.id', 'rules.severity')
+    end
```
Comment (skateman) [inline app/graphql/types/system.rb:79]: There's way too many joins here and the query seems to run separately for each system. This should be batch loaded...

**Review by skateman — CHANGES_REQUESTED** [review]

File: app/graphql/types/system.rb [hunk 5]
```diff
@@ -47,7 +47,12 @@ class System < Types::BaseObject
       argument :profile_id, String, 'Filter results by profile ID',
                required: false
     end
-
+    field :rules_failed_severity, [::Types::Rule], null: false do
+      argument :policy_id, ID, 'Filter results by system ID',
+               required: false
```
Comment (skateman) [inline app/graphql/types/system.rb:52]:
```suggestion
               required: true
```
maybe?

File: app/graphql/types/system.rb [hunk 6]
```diff
@@ -47,7 +47,12 @@ class System < Types::BaseObject
...
+      argument :severity, [String], 'Severity levels of failed rules',
+               required: false
```
Comment (skateman) [inline app/graphql/types/system.rb:54]: Same here, isn't it required? Or what happens if the severity is unset with the code below?

File: app/graphql/schema.graphql [hunk 7]
```diff
@@ -695,6 +695,17 @@ type System implements Node {
     """
     profileId: String
   ): Int!
+  rulesFailedBySeverity(
```
Comment (bastilian) [inline app/graphql/schema.graphql:698]: Ideally filtering parameters would be passed via the `filter` parameter, like with OS, or name, etc. and used by scoped search:
Screenshot 2023-04-19 at 13 43 13

File: app/graphql/types/system.rb [hunk 8]
```diff
@@ -64,6 +69,12 @@ def rules_failed(args = {})
       end
     end
 
+    def rules_failed_by_severity(args = {})
+      ::RecordLoader.for(::Host).load(object.id).then do |host|
+        host.failed_rules.reject { |rule| args[:severity].exclude?(rule.severity) }
```
Comment (skateman) [inline app/graphql/types/system.rb:74]: This is still causing an N+1, you need to have a second level of `CollectionLoader` that you load the `failed_rules` with.

**Comment by skateman** [comment]
Can you please provide a differential example query with the debug log output locally? Start with the GQL request that will be altered later in the frontend and see its timings and allocations. Then add the field you just created to that query and compare the two.

File: app/graphql/types/system.rb [hunk 9]
```diff
@@ -64,6 +69,14 @@ def rules_failed(args = {})
...
     end
 
+    def rules_failed_by_severity(args = {})
+      ::RecordLoader.for(::Host).load(object.id).then do |host|
+        ::CollectionLoader.for(::Host, :failed_rules).load(host).then do |failed|
+          failed.select { |rule| args[:severity].include?(rule.severity) }
+        end
+      end
+    end
+
```
Comment (skateman) [inline app/graphql/types/system.rb:79]: Not needed

File: app/graphql/resolvers/concerns/collection.rb [hunk 10]
```diff
@@ -22,6 +22,9 @@ module Collection
 
       def resolve(**kwargs)
         filters = filter_list(**kwargs)
+        puts "\u001b[31;1m◉\u001b[0m /home/rblanco/devel/compliance-backend/app/graphql/resolvers/concerns/collection.rb"
+        puts "kwargs: #{kwargs}"
+        puts "-" * 40
```
Comment (skateman) [inline app/graphql/resolvers/concerns/collection.rb:27]: Not needed.

File: app/graphql/types/system.rb [hunk 11]
```diff
@@ -47,7 +47,12 @@ class System < Types::BaseObject
       argument :profile_id, String, 'Filter results by profile ID',
                required: false
     end
-
+    field :rules_failed_by_severity, [::Types::Rule], null: false do
+      argument :policy_id, ID, 'Filter results by system ID',
+               required: true
+      argument :severity, [String], 'Severity levels of failed rules',
+               required: true
+    end
```
Comment (skateman) [inline app/graphql/types/system.rb:55]: Not needed

File: app/models/host.rb [hunk 12]
```diff
@@ -66,6 +66,8 @@ class Host < ApplicationRecord
 
   has_many :test_result_profiles, -> { distinct },
            through: :test_results, source: :profile
+  has_many :failed_rules, -> { ::RuleResult.failed },
+           through: :rule_results, source: :rule
```
Comment (skateman) [inline app/models/host.rb:70]: Not needed

File: app/models/concerns/host_searching.rb [hunk 13]
```diff
@@ -142,6 +145,17 @@ def filter_by_compliance_score(_filter, operator, score)
       { conditions: "hosts.id IN(#{hosts.to_sql})" }
     end
 
+    def filter_by_failed_rules_severity(_filter, operator, value)
+      values = value.split(',').map(&:strip)
+      profiles = RequestStore.store['scoped_search_context_profiles']
+      raise ScopedSearch::QueryNotSupported if profiles.nil?
+      # FIXME: duplicated on #filter_by_ssg_version
+      where = { rule: { severity: values } }
+      base = ::RuleResult.latest(profiles.compact.first.policy_id).joins(:rule).distinct.select(:host_id)
```
Comment (skateman) [inline app/models/concerns/host_searching.rb:154]: I was thinking about better solutions, but I guess this is the best one right now...

File: test/graphql/queries/system_query_test.rb [hunk 14]
```diff
@@ -261,6 +260,212 @@ class SystemQueryTest < ActiveSupport::TestCase
       assert_equal result.first['node']['id'], @host2.id
     end
 
+    should 'filter hosts by failed severe rules' do
```
Comment (skateman) [inline test/graphql/queries/system_query_test.rb:263]: Please split this into 2 separate tests.

File: test/graphql/queries/system_query_test.rb [hunk 15]
```diff
@@ -7,7 +7,6 @@ class SystemQueryTest < ActiveSupport::TestCase
     PolicyHost.any_instance.stubs(:host_supported?).returns(true)
     @user = FactoryBot.create(:user)
     @host1 = FactoryBot.create(:host, org_id: @user.account.org_id)
-
```
Comment (skateman) [inline test/graphql/queries/system_query_test.rb:10]: Unrelated change, not in this PR please

**Review by skateman — CHANGES_REQUESTED** [review]

File: app/models/concerns/host_searching.rb (hunk 13 above)
Comment (romanblanco) [reply app/models/concerns/host_searching.rb:154]: There's `FIXME:` to remind me to extract this into a method

Comment (skateman) [reply app/models/concerns/host_searching.rb:154]: I meant the `compact.first.policy_id` part

**Review by skateman — CHANGES_REQUESTED** [review]

File: test/graphql/queries/system_query_test.rb [hunk 16]
```diff
@@ -261,6 +261,182 @@ class SystemQueryTest < ActiveSupport::TestCase
...
+                id
+              }
+            }
+          }
+        }
+      GRAPHQL
+
+      filter = <<-FILTER
+        (policy_id = #{@policy.id}) and
+        (failed_rules_with_severity ^ (high low))
```
Comment (skateman) [inline test/graphql/queries/system_query_test.rb:425]: @bastilian we are abusing a little how scoped_search works :disappointed: so the order of the filtering is important, the `policy_id` has to go before the `failed_rules_with_severity`

Comment (bastilian) [reply test/graphql/queries/system_query_test.rb:425]: 👍 I'll try to make a note of that in the frontend.

**Comment by sthirugn** [comment]
awesome!
